src/Conversion/KrnlToLLVM/KrnlToLLVM.cpp

```diff
@@ -1895,6 +1849,202 @@ void mlir::checkConstantOutputs(
+    llvm::StringRef signature = sigAttr.getValue();
+    auto splitSig = signature.split('@');
+    llvm::StringRef inSig = splitSig.first;
+    llvm::StringRef outSig = splitSig.second;
+    inSignatures.emplace_back(inSig.str());
+    outSignatures.emplace_back(outSig.str());
+
+    return WalkResult::advance();
+  });
+  // When there is only a single entry point function in a model, use
+  // DEFAULT_DYN_ENTRY_POINT.
+  if (entryPointNames.size() == 1) {
+    entryPointNames[0] = DEFAULT_DYN_ENTRY_POINT;
+    entryPointNames[0].push_back('\0'); // null terminate the string.
+  }
+}
+
+void mlir::genSignatureFunction(ModuleOp module,
+    ArrayRef<std::string> entryPointNames, ArrayRef<std::string> inSignatures,
+    ArrayRef<std::string> outSignatures) {
+  MLIRContext *context = module.getContext();
+  Location loc = module.getLoc();
+  OpBuilder b(context);
+
+  // Common information.
+  Type i8Type = IntegerType::get(context, 8);
+  Type i32Type = IntegerType::get(context, 32);
+  Type i64Type = IntegerType::get(context, 64);
+  Type opaquePtrTy = LLVM::LLVMPointerType::get(i8Type);
+  Type opaquePtrPtrTy = LLVM::LLVMPointerType::get(opaquePtrTy);
+  IntegerAttr zeroI32Attr = b.getI32IntegerAttr(0);
+  IntegerAttr zeroI64Attr = b.getI64IntegerAttr(0);
+  IntegerAttr oneI64Attr = b.getI64IntegerAttr(1);
+
+  uint64_t numOfEntryPoints = entryPointNames.size();
+
+  // For each entry point name, emit three global constants to store the entry
+  // point name and input/output signatures. For the i-th entry point, these
+  // constants are named as follows:
+  // - Entry point name: `_entry_point_i`.
+  // - Input signature: `_entry_point_i_in_sig`.
+  // - Output signature: `_entry_point_i_out_sig`.
+  OpBuilder::InsertionGuard guard(b);
+  b.setInsertionPointToStart(module.getBody());
+  SmallVector<LLVM::GlobalOp, 2> entryOps, inSigOps, outSigOps;
+  for (uint64_t i = 0; i < numOfEntryPoints; ++i) {
+    // Global constants for entry point names prefixed by `run_`.
+    mlir::StringAttr entryAttr =
+        mlir::StringAttr::get(context, entryPointNames[i]);
+    Type entryArrayType =
+        LLVM::LLVMArrayType::get(i8Type, entryPointNames[i].size());
+    std::string entryVarName = "_entry_point_" + std::to_string(i);
+    LLVM::GlobalOp entryOp = b.create<LLVM::GlobalOp>(loc, entryArrayType,
+        /*isConstant=*/true, LLVM::Linkage::External, entryVarName, entryAttr);
+    entryOps.emplace_back(entryOp);
+
+    // Global constants for input signatures.
+    mlir::StringAttr inSigAttr =
+        mlir::StringAttr::get(context, inSignatures[i]);
+    Type inSigArrayType =
+        LLVM::LLVMArrayType::get(i8Type, inSignatures[i].size());
+    std::string inSigVarName = entryVarName + "_in_sig";
+    LLVM::GlobalOp inSigOp = b.create<LLVM::GlobalOp>(loc, inSigArrayType,
+        /*isConstant=*/true, LLVM::Linkage::External, inSigVarName, inSigAttr);
+    inSigOps.emplace_back(inSigOp);
```

Comment:
This code section looks similar to the one above (lines 1918-1925) and below (lines 1938-1946). Suggest to create a lambda or a templated function to avoid code duplication.